pkg/capabilities/validate.go

```diff
@@ -0,0 +1,210 @@
+	type CombinedWithoutInput struct {
+		Config Config `json:"config"`
+		Output Output `json:"output"`
+	}
+	var combinedWithoutInput = CombinedWithoutInput{}
+
+	var config interface{} = combined.Config
+	var input interface{} = combined.Input
+	var output interface{} = combined.Output
+	if config == nil {
+		return "", errors.New("config is nil, please provide a config type")
+	}
+	if output == nil {
+		return "", errors.New("output is nil, please provide an output type")
+	}
+
+	// we allow input to be nil, since triggers do not have inputs
+	if input == nil {
+		return schemaWith(v.SchemaReflector, combinedWithoutInput, v.schemas, "root", v.Info)
+	}
+
+	// print values of combined
+	return schemaWith(v.SchemaReflector, combined, v.schemas, "root", v.Info)
+}
+func (v *Validator[Config, Input, Output]) ConfigSchema() (string, error) {
+	return schemaWith(v.ConfigReflector, v.configType, v.schemas, "config", v.Info)
+}
+func (v *Validator[Config, Input, Output]) InputSchema() (string, error) {
+	return schemaWith(v.InputReflector, v.inputType, v.schemas, "input", v.Info)
+}
+func (v *Validator[Config, Input, Output]) OutputSchema() (string, error) {
+	return schemaWith(v.OutputReflector, v.outputType, v.schemas, "output", v.Info)
+}
+
+func (v *Validator[Config, Input, Output]) ValidateInput(input *values.Map) (*Input, error) {
+	inputSchema, err := v.InputSchema()
+	if err != nil {
+		return nil, errors.Join(errors.New("validation error while validating input"), err)
+	}
+
+	return validateAgainstSchema[Input](input, inputSchema)
+}
+func (v *Validator[Config, Input, Output]) ValidateConfig(config *values.Map) (*Config, error) {
+	configSchema, err := v.ConfigSchema()
+	if err != nil {
+		return nil, errors.Join(errors.New("validation error while validating config"), err)
+	}
+
+	return validateAgainstSchema[Config](config, configSchema)
+}
+func (v *Validator[Config, Input, Output]) ValidateOutput(output *values.Map) (*Output, error) {
+	outputSchema, err := v.OutputSchema()
+	if err != nil {
+		return nil, errors.Join(errors.New("validation error while validating output"), err)
+	}
+
+	return validateAgainstSchema[Output](output, outputSchema)
+}
+
+func validateAgainstSchema[DecodedValue any](value *values.Map, schema string) (*DecodedValue, error) {
+	jsonSchema, err := jsonvalidate.CompileString(uriPrefix, schema)
+	if err != nil {
+		return nil, err
+	}
+
+	raw, err := value.Unwrap()
+	if err != nil {
+		return nil, err
+	}
+
+	err = jsonSchema.Validate(raw)
+	if err != nil {
+		return nil, err
+	}
+
+	var decodedValue *DecodedValue
+	decoder, err := mapstructure.NewDecoder(&mapstructure.DecoderConfig{
+		DecodeHook: sliceToListHook,
+		Result:     &decodedValue,
+	})
+	if err != nil {
+		return nil, err
+	}
+	err = decoder.Decode(raw)
+	return decodedValue, err
+}
+
+func schemaWith(reflector *jsonschema.Reflector, schemaType any, schemaCache map[string]string, key string, info CapabilityInfo) (string, error) {
+	if reflector == nil {
```

Comment:
@HenryNguyen5 Can you avoid this by making reflector a `jsonschema.Reflector` to guarantee it's not nil.

Reply:
Would I just deref all of the args to it then?

Reply:
Yep, at the point at which this is called I'd expect you to be guaranteed of having a reflector